### PR TITLE
crystalline: init at 0.9.0

### LINKED ISSUE #129002: crystalline
**Project description**
Crystalline is an implementation of the Language Server Protocol written in and for the Crystal Language. It aims to provide limited language features (like go-to, autocompletion, syntax and semantic checking) and ease development with any compatible code editor.

**Metadata**

* homepage URL: https://github.com/elbywan/crystalline
* source URL: https://github.com/elbywan/crystalline
* license: mit
* platforms: linux, darwin

### PR DESCRIPTION
**Pull request by donovanglover**
###### Description of changes

Closes #129002

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/language-servers/crystalline/default.nix [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+}:
+
+crystal.buildCrystalPackage rec {
+  pname = "crystalline";
+  version = "0.9.0";
+
+  src = fetchFromGitHub {
+    owner = "elbywan";
+    repo = "crystalline";
+    rev = "v${version}";
```
Comment (Janik-Haag) [inline pkgs/development/tools/language-servers/crystalline/default.nix:16]: you should do and remove the rec:
```suggestion
    rev = "v${finalAttrs.version}";
```

for more information see: 
    History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
    Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
    Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

**Review by Janik-Haag — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/language-servers/crystalline/default.nix [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    wrapProgram "$out/bin/crystalline" --prefix PATH : '${
+      lib.makeBinPath [llvmPackages.llvm.dev]
+    }'
+  '';
+
+  meta = with lib; {
+    description = "A Language Server Protocol implementation for Crystal";
+    homepage = "https://github.com/elbywan/crystalline";
+    license = licenses.mit;
+    maintainers = with maintainers; [ donovanglover ];
```
[the remark this answers is not in the document]
Comment (pbek) [reply pkgs/development/tools/language-servers/crystalline/default.nix:43]: But I got many other build errors too. 😅

[the remark this answers is not in the document]
Comment (Janik-Haag) [reply pkgs/development/tools/language-servers/crystalline/default.nix:43]: Builds fine when doing:
```bash
gh pr checkout 238775
nix-build -A crystalline
```

[the remark this answers is not in the document]
Comment (pbek) [reply pkgs/development/tools/language-servers/crystalline/default.nix:43]: That's odd... I wonder why nixpkgs-review was tripping up for me.

[the remark this answers is not in the document]
Comment (Janik-Haag) [reply pkgs/development/tools/language-servers/crystalline/default.nix:43]: it does that sometimes also failes to build if for example upstream allready has the package.

File: pkgs/development/tools/language-servers/crystalline/default.nix (hunk 1 above)
Comment (donovanglover) [reply pkgs/development/tools/language-servers/crystalline/default.nix:19]: I wasn't able to figure out how to use `finalAttrs` with `crystal.buildCrystalPackage` so I used let / in.

File: pkgs/development/tools/language-servers/crystalline/default.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (donovanglover) [reply pkgs/development/tools/language-servers/crystalline/default.nix:46]: Fixed. `nixpkgs-review` seems very useful!

[the remark this answers is not in the document]
Comment (pbek) [reply pkgs/development/tools/language-servers/crystalline/default.nix:46]: Awesome 👍🏻😁

File: pkgs/development/tools/language-servers/crystalline/default.nix (hunk 1 above)
Comment (Janik-Haag) [reply pkgs/development/tools/language-servers/crystalline/default.nix:19]: Oh `finalAttrs` probably isn't implemented in `buildCrystalPackage` my mistake.  Using the let binding is fine, you can also go back to `rec` then it's fine for me either way. Sorry :see_no_evil:

**Review by Janik-Haag — APPROVED** [review]

File: pkgs/development/tools/language-servers/crystalline/default.nix (hunk 1 above)
Comment (donovanglover) [reply pkgs/development/tools/language-servers/crystalline/default.nix:19]: Thanks for the clarification. Will keep let/in then. Looking forward to the merge! :wink:

**Review by pbek — APPROVED** [review]
LGTM

**Pull request merged**